lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,160 @@
+"""
+
+import logging
+from lxml import etree
+from lxml.builder import ElementMaker
+import requests
+import requests_oauthlib
+import uuid
+
+from lti_provider.models import GradedAssignment, OutcomeService
+
+log = logging.getLogger("edx.lti_provider")
+
+
+def store_outcome_parameters(request_params, user, lti_consumer):
+    """
+    Determine whether a set of LTI launch parameters contains information about
+    an expected score, and if so create a GradedAssignment record. Create a new
+    OutcomeService record if none exists for the tool consumer, and update any
+    incomplete record with additional data if it is available.
+    """
+    result_id = request_params.get('lis_result_sourcedid', None)
+
+    # We're only interested in requests that include a lis_result_sourcedid
+    # parameter. An LTI consumer that does not send that parameter does not
+    # expect scoring updates for that particular request.
+    if result_id:
+        result_service = request_params.get('lis_outcome_service_url', None)
+        if not result_service:
+            # TODO: There may be a way to recover from this error; if we know
+            # the LTI consumer that the request comes from then we may be able
+            # to figure out the result service URL. As it stands, though, this
+            # is a badly-formed LTI request
+            log.warn(
+                "Outcome Service: lis_outcome_service_url parameter missing "
+                "from scored assignment; we will be unable to return a score. "
+                "Request parameters: %s",
+                request_params
+            )
+            return
+
+        # Both usage and course ID parameters are supplied in the LTI launch URL
+        usage_key = request_params['usage_key']
+        course_key = request_params['course_key']
+
+        # Create a record of the outcome service if necessary
+        outcomes, __ = OutcomeService.objects.get_or_create(
+            lis_outcome_service_url=result_service,
+            lti_consumer=lti_consumer
+        )
+
+        GradedAssignment.objects.get_or_create(
+            lis_result_sourcedid=result_id,
+            course_key=course_key,
+            usage_key=usage_key,
+            user=user,
+            outcome_service=outcomes
+        )
+
+
+def generate_replace_result_xml(result_sourcedid, score):
+    """
+    Create the XML document that contains the new score to be sent to the LTI
+    consumer. The format of this message is defined in the LTI 1.1 spec.
+    """
+    # Pylint doesn't recognize members in the LXML module
+    # pylint: disable=no-member
+    elem = ElementMaker(nsmap={None: 'http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0'})
+    xml = elem.imsx_POXEnvelopeRequest(
+        elem.imsx_POXHeader(
+            elem.imsx_POXRequestHeaderInfo(
+                elem.imsx_version('V1.0'),
+                elem.imsx_messageIdentifier(str(uuid.uuid4()))
+            ),
+            elem.imsx_POXBody(
```

Comment:
I noticed that you had tests to verify the generated XML -- is there a reason that test didn't catch it?